core/include/keyman/keyman_core_api.h

```diff
@@ -310,6 +310,38 @@ km_core_context_items_to_utf8(km_core_context_item const *item,
                               char *buf,
                               size_t *buf_size);
 
+/*
+```
+### `km_core_context_items_to_utf32`
+##### Description:
+Convert a context item array into a UTF-32 encoded string placing it into
+the supplied buffer of specified size, and return the number of codepoints
+actually used in the conversion. If null is passed as the buffer the
+number codepoints required is returned. This will strip markers from the
+context during the conversion.
+##### Return status:
+- `KM_CORE_STATUS_OK`: On success.
+- `KM_CORE_STATUS_INVALID_ARGUMENT`: If non-optional parameters are null.
+- `KM_CORE_STATUS_INSUFFICENT_BUFFER`: If the buffer is not large enough.
+  `buf_size` will contain the space required. The contents of the buffer are
+  undefined.
+##### Parameters:
+- __context_items__: A pointer to the start of an array `km_core_context_item`.
+    Must be terminated with a type of `KM_CORE_CT_END`.
+- __buf__: A pointer to the buffer to place the UTF-32 string into.
+    May be null to request size calculation.
+- __buf_size__: a pointer to the result variable:
+    The size of the supplied buffer in codepoints if `buf` is given.
+    On return will be the size required if `buf` is null.
```

Comment:
I realize that this if following the pattern of the UTF8 function, but I think it would be good to have a separate function to get the required size rather than using a null buf to request that. Without it, that existence of that behavior may be hidden in this comment.

Reply:
so ICU's idiom is to return the equivalent of `KM_CORE_STATUS_INSUFFICENT_BUFFER` if there isn't enough space, even if a buffer size was originally given.  And then, if you pass in 0 for the length ("pure preflight") then only calculation is done.

The 'mixed' use case could be where you preallocate a buffer of a certain size that should work for most cases, and then for a bad case you (re)allocate and try again.

Reply:
What is the return status when `nullptr` is passed in? 

I think `buf_size` should be 0 if it's for size calculation.

Reply:
The using null to request size is a pretty established pattern (not just this core), I am happy to add a `get_buffer_size` call but would suggest leaving this null_buf to behaviour with this function also.  In fact the the `get_buffer_size` could just call this function with the null_buff.

Reply:
Given we already have two other functions that follow this pattern, I think it makes more sense to be consistent -- if we think this buffer sizing pattern is bad, we should fix all three functions rather than introduce a new pattern for just one of them. But that should be a separate change and not in this PR.

Reply:
Ok, I was wondering whether that was a common pattern. I really dislike it when null is a magic value, but it doesn't make sense to change the rules, especially for an internal API.